File: 2509-minimize-xor/minimize-xor.ts
function minimizeXor(num1: number, num2: number): number {
    let setbit = 0;
    while(num2 > 0) {
        setbit += num2 & 1;
        num2 >>= 1;
    }
    const ori = num1.toString(2).split('');
    const next = new Array(ori.length).fill('0');
    for(let i = 0; i < ori.length && setbit > 0; i++) {
        if(ori[i] === '1') {
            next[i] = '1'
            setbit--
        }
    }
    for(let i = ori.length - 1; i >= 0 && setbit > 0; i--) {
        if(ori[i] === '0') {
            next[i] = '1'
            setbit--
        }
    }
    if(setbit > 0) next.unshift('1'.repeat(setbit))
    return parseInt(next.join(''), 2)
};

File: 2509-minimize-xor/minimize-xor.rs
impl Solution {
    pub fn minimize_xor(num1: i32, num2: i32) -> i32 {
        let mut setbit = 0;
        let mut _num2 = num2;
        let mut res = 0;
        while _num2 > 0 {
            if _num2 & 1 == 1 {
                setbit += 1;
            }
            _num2 >>= 1;
        }
        let mut cur = (2_i32).pow((num1 as f64).log2().floor() as u32);
        while cur > 0 && setbit > 0 {
            if num1 & cur == cur {
                res += cur;
                setbit -= 1;
            }
            cur >>= 1;
        }
        cur = 1;
        while setbit > 0 {
            if num1 & cur == 0 {
                res += cur;
                setbit -= 1;
            }
            cur <<= 1;
        }
        res
    }
}